### PR TITLE
Fail fast

### PR DESCRIPTION
**Pull request by bcsaller**
This adds an exception plugins can raise that cleanly fails a test. Likewise non-zero exits from executable tasks  (which have fewer options than plugins) will now cleanly fail tests.

### CODE REVIEW

**Comment by petevg** [comment]
I am +1 on merging this, esp. seeing it used in @johnsca's branch.

**Comment by bcsaller** [comment]
so the current version has better handling around test fails and should allow a "gating" bool to be passed in the task yaml def which controls the higher level pass/fail behaviour

File: matrix/model.py [hunk 1]
```diff
@@ -211,7 +218,9 @@ def attr_filter(a, v):
                 rule.log.debug(stdout.decode('utf-8'))
             if stderr:
                 rule.log.debug(stderr.decode('utf-8'))
-            result = p.returncode is 0
+            result = p.returncode == 0
```
Comment (petevg) [inline matrix/model.py]: `p.returncode is 0` seems more idiomatic. But this code works fine.

**Comment by petevg** [comment]
+1 on updates.

**Pull request merged**